File: server/java/src/main/java/com/onlinepayments/example/api/validators/common/card/CardValidator.java
package com.onlinepayments.example.api.validators.common.card;

import com.onlinepayments.example.api.exceptions.ApiValidationException;
import com.onlinepayments.example.service.domain.payments.Card;

import java.time.YearMonth;

import static com.onlinepayments.example.api.validators.BaseValidator.isNullOrEmpty;

public class CardValidator {
    public static void validateCard(Card card) {
        if (card == null) {
            throw new ApiValidationException("The Card object is required.");
        }

        if (isNullOrEmpty(card.number())) {
            throw new ApiValidationException("The Card.Number field is required.");
        }

        if (!card.number().chars().allMatch(Character::isDigit)) {
            throw new ApiValidationException("The Card.Number field must contain only digits.");
        }

        if (card.number().length() > 19) {
            throw new ApiValidationException(" The Card.Number field must be shorter than 20 characters.");
        }

        if (isNullOrEmpty(card.holderName())) {
            throw new ApiValidationException("The Card.HolderName field is required.");
        }

        if (isNullOrEmpty(card.verificationCode())) {
            throw new ApiValidationException("The Card.VerificationCode field is required.");
        }

        if (card.verificationCode().length() < 3 || card.verificationCode().length() > 4 || !card.verificationCode().chars().allMatch(Character::isDigit)) {
            throw new ApiValidationException("The Card.VerificationCode must be 3 or 4 digits long.");
        }

        if (!card.verificationCode().chars().allMatch(Character::isDigit)) {
            throw new ApiValidationException("The Card.VerificationCode for CREDIT_CARD payments must contain only digits.");
        }

        if (isNullOrEmpty(card.expiryMonth())) {
            throw new ApiValidationException("The Card.ExpiryMonth field is required.");
        }

        if (isNullOrEmpty(card.expiryYear())) {
            throw new ApiValidationException("The Card.ExpiryYear field is required.");
        }

        validateCardExpiryDate(card.expiryMonth(), card.expiryYear());
    }

    private static void validateCardExpiryDate(String expiryMonth, String expiryYear) {
        if (expiryMonth == null || expiryMonth.isBlank()) {
            throw new ApiValidationException("The Card.ExpiryMonth is required.");
        }

        if (expiryMonth.length() != 2) {
            throw new ApiValidationException("The Card.ExpiryMonth field must be 2 digits long.");
        }

        int month;
        try {
            month = Integer.parseInt(expiryMonth);
        } catch (NumberFormatException e) {
            throw new ApiValidationException("The Card.ExpiryMonth field must be a number.");
        }

        if (month < 1 || month > 12) {
            throw new ApiValidationException("The Card.ExpiryMonth must be a number between 01 and 12.");
        }

        if (expiryYear == null || expiryYear.isBlank()) {
            throw new ApiValidationException("The Card.ExpiryYear is required.");
        }

        if (expiryYear.length() != 4) {
            throw new ApiValidationException("The Card.ExpiryYear field must be 4 digits long.");
        }

        int year;
        try {
            year = Integer.parseInt(expiryYear);
        } catch (NumberFormatException e) {
            throw new ApiValidationException("The Card.ExpiryYear field must be a number.");
        }

        YearMonth expiry = YearMonth.of(year, month);
        if (!expiry.isAfter(YearMonth.now())) {
            throw new ApiValidationException("The card expiry date must be in the future.");
        }
    }
}

File: server/python/app/presentation/validators/common/card/card_validator.py
import re
import calendar
from datetime import date
from app.application.domain.payments.card import Card

class CardValidator:

    def validate(self, card: Card, prefix: str = "Card") -> list[str]:
        errors = []

        if not card:
            errors.append(f"The {prefix} object is required.")
            
            return errors

        number = (card.number or "").strip()
        holder_name = (card.holder_name or "").strip()
        verification_code = (card.verification_code or "").strip()
        expiry_month = str(card.expiry_month or "").strip()
        expiry_year = str(card.expiry_year or "").strip()

        if not number:
            errors.append(f"The {prefix}.Number field is required.")
        elif not re.match(r'^\d+$', number):
            errors.append(f"The {prefix}.Number field must be a valid number.")
        elif len(number) > 19:
            errors.append(f"The {prefix}.Number field must be shorter than 20 characters.")

        if not holder_name:
            errors.append(f"The {prefix}.HolderName field is required.")

        if not verification_code:
            errors.append(f"The {prefix}.VerificationCode field is required.")
        elif not re.match(r'^\d{3,4}$', verification_code):
            errors.append(f"The {prefix}.VerificationCode field must be 3 or 4 digits long.")

        if not expiry_month:
            errors.append(f"The {prefix}.ExpiryMonth field is required.")
        elif not re.match(r'^\d+$', expiry_month):
            errors.append(f"The {prefix}.ExpiryMonth field must be a number.")
        elif len(expiry_month) != 2:
            errors.append(f"The {prefix}.ExpiryMonth field must be 2 digits long.")
        elif not (1 <= int(expiry_month) <= 12):
            errors.append(f"The {prefix}.ExpiryMonth field must be a number between 01 and 12.")

        if not expiry_year:
            errors.append(f"The {prefix}.ExpiryYear field is required.")
        elif not re.match(r'^\d+$', expiry_year):
            errors.append(f"The {prefix}.ExpiryYear field must be a number.")
        elif len(expiry_year) != 4:
            errors.append(f"The {prefix}.ExpiryYear field must be 4 digits long.")

        if not errors and expiry_month and expiry_year:
            if re.match(r'^\d+$', expiry_month) and re.match(r'^\d+$', expiry_year):
                if len(expiry_month) == 2 and len(expiry_year) == 4:
                    m = int(expiry_month)
                    y = int(expiry_year)
                    if 1 <= m <= 12:
                        try:
                            last_day = calendar.monthrange(y, m)[1]
                            expiry_date = date(y, m, last_day)
                            if expiry_date <= date.today():
                                errors.append("The card expiry date must be in the future.")
                        except ValueError:
                            errors.append("The card expiry date is invalid.")

        return errors